dsh/dsh.1
.\" $Id: dsh.1,v 1.21 2007/01/10 20:36:11 garbled Exp $
.\" Copyright (c) 1998, 1999, 2000
.\"	Tim Rightnour.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgment:
.\"	This product includes software developed by Tim Rightnour.
.\" 4. The name of Tim Rightnour may not be used to endorse or promote 
.\"    products derived from this software without specific prior written 
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY TIM RIGHTNOUR ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL TIM RIGHTNOUR BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" The following requests are required for all man pages.
.Dd January 9, 2007
.Dt DSH 1
.Sh NAME
.Nm dsh
.Nd run a command on a cluster of machines
.Sh SYNOPSIS
.Nm
.Op Fl eiqtv
.Op Fl f Ar fanout
.Op Fl g Ar rungroup1,...,rungroupN
.Op Fl l Ar username
.Op Fl o Ar porttimeout
.Op Fl p Ar portnum
.Op Fl w Ar node1,...,nodeN
.Op Fl x Ar node1,...,nodeN
.Op Ar command ...
.Nm
.Op Fl eiqtv
.Op Fl f Ar fanout
.Op Fl g Ar rungroup1,...,rungroupN
.Op Fl l Ar username
.Op Fl o Ar porttimeout
.Op Fl p Ar portnum
.Op Fl w Ar node1,...,nodeN
.Op Fl x Ar node1,...,nodeN
.Fl s Ar scriptname
.Op Ar arguments ...
.Sh DESCRIPTION
The 
.Nm
utility can be used to run a command, or group of commands on a cluster of 
machines.  All commands are run in parallel, on the cluster.  Interrupt
signals will be sent to the remote host that is currently being displayed
to the user.
The following options are available:
.Bl -tag -width www
.It Fl e
Unless the
.Fl e
option is specified, stderr from remote commands will not be reported to the
user.
.It Fl f
If the
.Fl f
option is specified, followed by a number, it sets the fanout size of the 
cluster.  The fanout size is the number of nodes a command will run on in 
parallel at one time.  Thus a 80 node cluster, with a fanout size of 64, 
would run 64 nodes in parallel, then, when all have finished, it would 
execute the command on the last 16 nodes.  The fanout size defaults to 64. 
This option overrides the
.Ev FANOUT
environment variable.
.It Fl g
If the
.Fl g
option is specified, followed by a comma separated list of group names, the 
command will only be run on that group of nodes.  A node may be a part of 
more than one group if desired, however running without the
.Fl g
option will run the command on the same node as many times as it appears in the
file specified by the
.Ev CLUSTER
environment variable.  This option is silently ignored if used with the
.Fl w
option.
.It Fl i
The
.Fl i
option will list information about the current cluster, and command groupings. 
It will print out the current value of the fanout, and how many groups of 
machines there are within the cluster. It will also show you which command 
you are about to run, and your username if specified with the
.Fl l
option.
.It Fl l
If the
.Fl l
option is specified, followed by a username, the commands will be run under 
that userid on the remote machines.  Consideration must be taken for proper 
authentication, for this to work.
.It Fl o
The
.Fl o
option is used to set the timeout in seconds to be used when testing remote
connections.  The default is five seconds.
.It Fl p
The
.Fl p
option can be used to set the port number that testing should occur on when
testing remote connections.  The default behavior is to guess based on the
remote command name.
.It Fl q
The
.Fl q
option does not issue any commands, but displays information about the 
cluster, and the fanout groupings.
.It Fl s
The
.Fl s
option causes
.Nm
to copy a script to the remote machine, execute it once, and delete it, all
in a single operation.  The
.Fl s
option requires a script name, which will be copied to all remote machines and
executed.  You may also optionally specify any number of additional arguments
to the script on the command line.  The script will be placed in a temporary
directory under /tmp on the remote node, executed, and then the directory will
be recursively deleted.  Any executable can be used as the script, regardless
of programming language.  The script is copied with the tar command, preserving
permissions of the original.  The
.Fl s
option cannot be used with the standard mode of
.Nm
to run other commands, nor can it be used in interactive mode.
.It Fl t
The
.Fl t
option causes
.Nm
to attempt a connection test to each node prior to attempting to run the
remote command.  If the test fails for any reason, the remote command will
not be attempted.  This can be useful when clusterfiles have suffered bitrot
and some nodes no longer exist, or might be down for maintenance.  The default
timeout is 5 seconds.  The timeout can be changed with the
.Fl o
option.
.Nm
will attempt to guess the port number of the remote service based on your
.Ev RCMD_CMD
setting.  It knows about ssh and rsh.  If
.Nm
fails to guess your port correctly, you may use the
.Fl p
argument to set the remote port number.  If the
.Ev RCMD_TEST
environment variable exists, the testing will automatically take place.
.It Fl v
Prints the version of ClusterIt to the stdout, and exits.
.It Fl w
If the
.Fl w
option is specified, followed by a comma delimited list of machine names,
the
.Ar command
will be run on each node in the list.  Without this option,
.Nm
runs on the nodes listed in the file pointed to by the
.Ev CLUSTER
environment variable.
.It Fl x
The
.Fl x
option can be used to exclude specific nodes from the cluster.  The format 
is the same as the
.Fl w
option, a comma delimited list of machine names.  This option is silently 
ignored if used with the
.Fl w
option.
.El
.Sh ENVIRONMENT
.Nm
utilizes the following environment variables.
.Bl -tag -width "RCMD_TEST_TIMEOUT"
.It Ev CLUSTER
Contains a filename, which is a newline separated list of nodes
in the cluster.
.It Ev RCMD_CMD
Command to use to connect to remote machines.  The command chosen must
be able to connect with no password to the remote host.  Defaults to
.Ic rsh
.It Ev RCMD_CMD_ARGS
Arguments to pass to the remote shell command.  Defaults to none.
.It Ev RCMD_PORT
The port number used to test remote connections.  See the
.Fl p
flag.
.It Ev RCMD_TEST
When set,
.Nm
will automatically test all hosts before launching the remote command. See the
.Fl t
option for more information.
.It Ev RCMD_TEST_TIMEOUT
The timeout in seconds to use when testing for remote connections.
.It Ev RCMD_USER
The username to connect to remote machines as by default.
.It Ev FANOUT
When set, limits the maximum number of concurrent commands sent at once.  
This can be used to keep from overloading a small host when sending out 
commands in parallel.  Defaults to 64.  This environment setting can be 
overridden by the
.Fl f
option.
.El
.Sh FILES
The file pointed to by the
.Ev CLUSTER
environment variable has the following format:
.Bd -literal -offset indent
pollux
castor
GROUP:alpha
rigel
kent
GROUP:sparc
alshain
altair
LUMP:alphasparc
alpha
sparc
.Ed
.Pp
This example would have pollux and castor a member of no groups, rigel and
kent a member of group 'alpha', and alshain and altair a member of group 
.Sq sparc .
Note the format of the GROUP command, it is in all capital letters, followed
by a colon, and the group name.  There can be no spaces following the GROUP
command, or in the name of the group.
.Pp
There is also a LUMP command, which is identical in syntax to the GROUP
command.  This command allows you to create a named group of groups.  Each
member of the lump is the name of a group.  The LUMP command is terminated
by another LUMP or GROUP command, or the EOF marker.
.Pp
Any line beginning with a
.Sq #
symbol denotes a comment field, and the entire line will be ignored.
Note that a hash mark placed anywhere other than the first character
of a line, will be considered part of a valid hostname or command.
.Sh EXAMPLES
The command:
.Bd -literal -offset indent
dsh hostname
.Ed
.Pp
will display:
.Bd -literal -offset indent
pollux:	pollux
castor:	castor
.Ed
.Pp
if the file pointed to by
.Ev CLUSTER
contains:
.Bd -literal -offset indent
pollux
castor
.Ed
.Pp
The command:
.Bd -literal -offset indent
dsh -w hadar,rigel hostname
.Ed
.Pp
will display:
.Bd -literal -offset indent
hadar:	hadar
rigel:	rigel
.Ed
.Pp
The command:
.Bd -literal -offset indent
dsh -w hadar,rigel -s /bin/date
.Ed
.Pp
Will copy /bin/date to /tmp/dsh.$$ on hadar and rigel and execute it on 
each node, displaying the date and time on each remote machine, assuming
that the /bin/date you copied is a valid binary for the remote end.
.Sh DIAGNOSTICS
Exit status is 0 on success, 1 if an error occurs.
.Sh SEE ALSO
.Xr dshbak 1 ,
.Xr pcp 1 ,
.Xr pdf 1 ,
.Xr prm 1 ,
.Xr rsh 1 ,
.Xr tar 1 ,
.Xr kerberos 3 ,
.Xr hosts.equiv 5 ,
.Xr rhosts 5
.Sh HISTORY
The
.Nm
command appeared in clusterit 1.0. It is based on the
.Nm
command in IBM PSSP.
.Sh AUTHOR
.Nm Dsh
was written by Tim Rightnour.
.Sh BUGS
Solaris 2.5.1 has a maximum of 256 open file descriptors.  This means
that
.Nm
will fail on a fanout size greater than about 32-40 nodes.


CHANGES
$Id: CHANGES,v 1.18 2007/01/04 18:59:06 garbled Exp $

Add -s option to dsh.  This adds the ability to copy a script to a remote
machine, execute it, and delete it, all in a single operation.

Add RCP_CMD_ARGS processing to pcp.

* Release 2.4 *

Fix jsd to work properly on Linux.

Fix stdin command mode on all the utilities.

Make jsh and jsd more robust.  Keep jsd from permanently losing it's nodes.

Fix pdf to work on Linux.

Add a -v option to all programs to tell the user what version of
ClusterIt they are running.  CVS will allways have _BETA tagged on the end,
and it will be removed at release.

Add RVT_CMD_ARGS to dvt, to allow passing of args to rvt

Incorporate patches from debian xvt package into rvt:    
Fix xvt manpage typos
Add -rv (reverse-video) option 
Better colormap handling
All other patches were either irrelevant (makefile fixes, etc) or allready
incorporated.

Fix pdf to work on linux hosts

Replace most instances of sprintf with snprintf. Did not touch rvt, as I would
prefer to keep that as unmodified as possible, and most of them would not
be easy to fix anyhow.

Replace most instances of malloc with calloc.   Mostly to prevent any
unanticipated wierdness, and a few of the mallocs were immediately followed
by memset commands anyhow.

Add RCMD_CMD_ARGS to most of the programs. This makes it easier to do
things like run "ssh -4" as your RCMD_CMD.

Stop allocating the command to be run on the remote end as buf[MAXBUF],
instead compute it at runtime and dynamically allocate it to prevent
possible overflow on a huge command.

Fix a bug noted by Jot Powers, where trailing whitespace on a GROUP or LUMP
entry would cause dsh -g to not match it.

Fix a bug noted by John Bales  where linux machines often recieved truncated 
output from the child ssh/rsh process.

Fix bug where all the programs would mangle argv[0] for ps.

* Release 2.3.1 *

Fix a bug noted by James F Wilkus on OpenBSD where dsh would segv when
reading the clusterfile.

Convert seq over to rseq fully, not just for linux anymore.  Sigh.

Fix the clusterit.spec file to work properly.  Tested on redhat

* Release 2.3 *

Made all remote commands use poll instead of a while loop to read the incoming
stream of data from stdout and stderr.  This fixes bug where linux ssh will
hang up.

Fixed linux command line argument processing.

Changed the -l <username> processing to invoke RCMD_CMD with user@host rather
than -l user host.  This makes it much more reliable.

Added an option to dsh to test remote ports before connecting.  This can be
useful where a remote host might be down, and a long timeout would occur
waiting for a response.

Updated rvt to use a newer set of sources from xvt.  Fixes questionable
licensing in xvt, and makes it more reliable.

Converted makefiles to autoconf for ease of porting.  No more stupid patch
sets which were a terrible burden to maintain and generate.

ANSIfyed most of the code (other than rvt). Nobody uses KNF anymore.. sigh..

Incorporated many patches sent by users from the internet.  I have
unfortunately lost track of what patch came from who, but thier help is
greatly appreciated.  Most of these patches either fixed bugs, added stability
or fixed porting bugs.  The xvt conversion was based mainly on a patch sent
in by a user.  My thanks go out to those of you who have sent them in.

# changes since 2.2

* Release 2.2 *
Missing this data, sorry
* Release 2.0 *

Added dvt and rvt..  an interactive form of dsh with xterms.

Fixed a bug in jsd and dsh where it would lock up running ssh waiting for
input.

Add a jsd dir, and move jsd.c from dsh to the jsd dir.  Add nearly complete
jsh program, and jcommon.[ch].

Add new common subdir and move a tonnage of crap in there.  Mainly
common.c, common.h and strsep.c.

Fix horrid bug in dsh where dsh | more would lock terminals up.  Still
need to fix in seq, run, probably pcp.

Add <netinet/if_ether.h> to barrierd.c to shut AIX C up.

Lots of minor bugfixes, mostly dealing with failures to open files, opening
0 byte clusters, etc etc..

Pcp was still using it's own code to parse strings and groups.  Fully rewrote
the back-end internals to use the new linked-list stuff, and utilize common.c.
Pcp's code is now virtually identical to dsh, whereas before it had diverged
significantly.

Error handling has finally graduated to where I no longer feel that it
"needs work".

Fix a place where malloced data was strcatted to without being bzeroed.

Group, exclusion and -w nodes are now dynamicly generated arrays.  Removes
the MAX_GROUPS restrictions.

The nodelist and grouplists are no longer maintained in any way similar to
the original setup.  Nodes are not stored in a flat array of nodes, with a
corresponding flat array of group names.  Nodes are now stores in a linked
list of machines to run on.  Nodes which are not to be run on, are not
included in this list, so the back-end routines only need to execute code
on every node they are handed in the linked list.  This is all handled in
common.c now.  This removes the all compile time restrictions on MAXNODES,
allowing the program to scale much more efficiently.

# changes since 1.3
* Release 1.3 *

Break dsh, run and seq's common code into a common.c and common.h file,
so I don't forget to add options to run and seq when fixing dsh.  Also, fix
some locations where run and seq were out of sync slightly.

Fix a bug in pdf where if the -m option was used, filesystems at 100%
would not be displayed.

Handle ^C's correctly.  If a remote process is hung, and it is the current
process being read, you may ^C it, and the interrupt will be sent to the
remote process.  This would be just like hitting ^C on the remote end, 
and will only affect the process you are currently reading from.

Added ability to pcp multiple files and wildcards to remote hosts.

Made pcp and dsh compile on an ANSI compiler (IBM C, SUN C)

Make output from dsh look nicer, and format well.

# changes since 1.2
* Release 1.2 *


dsh/dsh.c
/* $Id: dsh.c,v 1.32 2007/01/11 20:19:21 garbled Exp $ */
/*
 * Copyright (c) 1998, 1999, 2000
 *	Tim Rightnour.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by Tim Rightnour.
 * 4. The name of Tim Rightnour may not be used to endorse or promote 
 *    products derived from this software without specific prior written 
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY TIM RIGHTNOUR ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL TIM RIGHTNOUR BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/time.h>
#include <sys/types.h>
#include <sys/resource.h>
#include <sys/wait.h>

#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <poll.h>
#include <libgen.h>

#include "../common/common.h"

#if !defined(lint) && defined(__NetBSD__)
__COPYRIGHT(
"@(#) Copyright (c) 1998, 1999, 2000\n\
        Tim Rightnour.  All rights reserved\n");
__RCSID("$Id: dsh.c,v 1.32 2007/01/11 20:19:21 garbled Exp $");
#endif /* not lint */

void do_command(char **argv, int fanout, char *username);
void sig_handler(int i);

/* globals */
int debug, errorflag, gotsigint, gotsigterm, exclusion, grouping;
int testflag, rshport, porttimeout, script;
node_t *nodelink;
group_t *grouplist;
char **rungroup;
int nrofrungroups;
char **lumplist;
pid_t currentchild;
char *progname;
char *scriptname;

/* 
 *  dsh is a cluster management tool based upon the IBM tool of the
 *  same name.  It allows a user, or system administrator to issue
 *  commands in paralell on a group of machines.
 */

int
main(int argc, char *argv[])
{
    extern char	*optarg;
    extern int	optind;
    extern char *version;

    int someflag, ch, fanout, showflag, fanflag;
    char *p, *group, *nodename, *username;
    char **exclude;
    struct rlimit limit;
    node_t *nodeptr;

    someflag = showflag = fanflag = 0;
    exclusion = debug = errorflag = 0;
    testflag = rshport = nrofrungroups = 0;
    porttimeout = 5; /* 5 seconds to port timeout */
    gotsigint = gotsigterm = grouping = 0;
    fanout = DEFAULT_FANOUT;
    nodename = NULL;
    username = NULL;
    group = NULL;
    nodeptr = NULL;
    nodelink = NULL;
    exclude = NULL;

    rungroup = calloc(GROUP_MALLOC, sizeof(char **));
    if (rungroup == NULL)
	bailout();

    progname = strdup(basename(argv[0]));

#if defined(__linux__)
    while ((ch = getopt(argc, argv, "+?deiqtf:g:l:o:p:s:vw:x:")) != -1)
#else
    while ((ch = getopt(argc, argv, "?deiqtf:g:l:o:p:s:vw:x:")) != -1)
#endif
	switch (ch) {
	case 'd':		/* we want to debug dsh (hidden)*/
	    debug = 1;
	    break;
	case 'e':		/* we want stderr to be printed */
	    errorflag = 1;
	    break;
	case 'i':		/* we want tons of extra info */
	    debug = 1;
	    break;
	case 'l':		/* invoke me as some other user */
	    username = strdup(optarg);
	    break;
	case 'q':		/* just show me some info and quit */
	    showflag = 1;
	    break;
	case 't':           /* test the nodes before connecting */
	    testflag = 1;
	    break;
	case 'p':           /* what is the rsh port number? */
	    rshport = atoi(optarg);
	    break;
	case 'o':               /* set the test timeout in seconds */
	    porttimeout = atoi(optarg);
	    break;
	case 's':	/* we want to push a script and run it */
	    script = 1;
	    scriptname = strdup(optarg);
	    break;
	case 'f':		/* set the fanout size */
	    fanout = atoi(optarg);
	    fanflag = 1;
	    break;
	case 'g':		/* pick a group to run on */
	    grouping = 1;
	    nrofrungroups = parse_gopt(optarg);
	    break;
	case 'v':
	    (void)printf("%s: %s\n", progname, version);
	    exit(EXIT_SUCCESS);
	    break;
	case 'x':		/* exclude nodes, w overrides this */
	    exclusion = 1;
	    exclude = parse_xopt(optarg);
	    break;
	case 'w':		/* perform operation on these nodes */
	    someflag = 1;
	    for (p = optarg; p != NULL; ) {
		nodename = (char *)strsep(&p, ",");
		if (nodename != NULL)
		    (void)nodealloc(nodename);
	    }
	    break;
	case '?':		/* you blew it */
	    (void)fprintf(stderr,
	        "usage:\n%s [-eiqtv] [-f fanout] [-p portnum] [-o timeout] "
		"[-g rungroup1,...,rungroupN]\n"
		"    [-l username] [-x node1,...,nodeN] [-w node1,..,nodeN] "
		"[command ...]\n", progname);
	    (void)fprintf(stderr,
		"%s [-eiqtv] [-f fanout] [-p portnum] [-o timeout] "
		"[-g rungroup1,...,rungroupN]\n"
		"    [-l username] [-x node1,...,nodeN] [-w node1,..,nodeN]\n"
	        "    -s scriptname [arguments ...]\n", progname);
	    return(EXIT_FAILURE);
	    /*NOTREACHED*/
	    break;
	default:
	    break;
	}

    /* Check for various environment variables and use them if they exist */
    if (!fanflag && getenv("FANOUT"))
        fanout = atoi(getenv("FANOUT"));
    if (!rshport && getenv("RCMD_PORT"))
	rshport = atoi(getenv("RCMD_PORT"));
    if (!testflag && getenv("RCMD_TEST"))
	testflag = 1;
    if (porttimeout == 5 && getenv("RCMD_TEST_TIMEOUT"))
	porttimeout = atoi(getenv("RCMD_TEST_TIMEOUT"));
    if (username == NULL && getenv("RCMD_USER"))
	username = strdup(getenv("RCMD_USER"));

    rshport = get_rshport(testflag, rshport, "RCMD_CMD");

    if (!someflag)
	parse_cluster(exclude);

    argc -= optind;
    argv += optind;
    if (showflag) {
	do_showcluster(fanout);
	return(EXIT_SUCCESS);
    }

    /*
     * set per-process limits for max descriptors, this avoids running
     * out of descriptors and the odd errors that come with that.
     */
    if (getrlimit(RLIMIT_NOFILE, &limit) != 0)
	bailout();
    if (limit.rlim_cur < fanout * 5) {
	limit.rlim_cur = fanout * 5;
	if (setrlimit(RLIMIT_NOFILE, &limit) != 0)
	    bailout();
    }

    do_command(argv, fanout, username);
    return(EXIT_SUCCESS);
}

/* 
 * Do the actual dirty work of the program, now that the arguments
 * have all been parsed out.
 */

void
do_command(char **argv, int fanout, char *username)
{
    struct sigaction signaler;
    FILE *fd, *fda, *in;
    char buf[MAXBUF], cbuf[MAXBUF], pipebuf[2048];
    int status, i, j, n, g, piping, pollret, nrofargs, arg, slen;
    size_t maxnodelen;
    char *p, **q, *command, **rsh, *cd, **cmd, *rshstring;
    char *scriptbase, *scriptdir;
    node_t *nodeptr, *nodehold;
    struct pollfd fds[2];

    maxnodelen = 0;
    j = i = 0;
    piping = 0;
    in = NULL;
    cd = pipebuf;

    if (debug) {
	if (username != NULL)
	    (void)printf("As User: %s\n", username);
	(void)printf("On nodes:\n");
    }
    for (nodeptr = nodelink; nodeptr; nodeptr = nodeptr->next) {
	if (strlen(nodeptr->name) > maxnodelen)
	    maxnodelen = strlen(nodeptr->name);
	if (debug) {
	    if (!(j % 4) && j > 0)
		(void)printf("\n");
	    (void)printf("%s\t", nodeptr->name);
	}
	j++;
    }

    i = j; /* side effect of above */
    j = i / fanout;
    if (i % fanout)
	j++; /* compute the # of rungroups */

    /* construct the command from the remains of argv */
    for (i=0, p=*argv, q=argv; p != NULL; p = *++q)
	i += (strlen(p)+1);
    command = (char *)calloc(i+1, sizeof(char));
    for (p = *argv; p != NULL; p = *++argv ) {
	strcat(command, p);
	strcat(command, " ");
    }
    if (debug) {
	(void)printf("\nDo Command: %s\n", command);
	(void)printf("Fanout: %d Groups:%d\n", fanout, j);
    }
    if (strcmp(command, "") == 0 && !script) {
	piping = 1;
	/* are we a terminal?  then go interactive! */
	if (isatty(STDIN_FILENO) && piping)
	    (void)printf("%s>", progname);
	in = fdopen(STDIN_FILENO, "r");
	/* start reading stuff from stdin and process */
	command = fgets(cbuf, sizeof(cbuf), in);
	if (command != NULL)
	    if (strcmp(command,"\n") == 0)
		command = NULL;
    } else {
	close(STDIN_FILENO); /* DAMN this bug took awhile to find */
	if (open("/dev/null", O_RDONLY, NULL) != 0)
	    bailout();
    }

    signaler.sa_handler = sig_handler;
    signaler.sa_flags = SA_RESTART;
    sigaction(SIGTERM, &signaler, NULL);
    sigaction(SIGINT, &signaler, NULL);
    sigaction(SIGALRM, &signaler, NULL);

    rsh = parse_rcmd("RCMD_CMD", "RCMD_CMD_ARGS", &nrofargs);
    rshstring = build_rshstring(rsh, nrofargs);
    
    /* begin the processing loop */
    g = 0;
    nodeptr = nodelink;
    while (command != NULL) {
	for (n=0; n <= j; n++) {
	    if (gotsigterm || gotsigint)
		exit(EXIT_FAILURE);
	    nodehold = nodeptr;
	    for (i=0; (i < fanout && nodeptr != NULL); i++) {
		g++;
		if (gotsigterm)
		    exit(EXIT_FAILURE);
		if (debug)
		    (void)printf("Working node: %d, fangroup %d,"
			" fanout part: %d\n", g, n, i);
/*
 * we set up pipes for each node, to prepare for the oncoming barrage of data.
 * Close on exec must be set here, otherwise children spawned after other
 * children, inherit the open file descriptors, and cause the pipes to remain
 * open forever.
 */
		if (pipe(nodeptr->out.fds) != 0)
		    bailout();
		if (pipe(nodeptr->err.fds) != 0)
		    bailout();
		if (fcntl(nodeptr->out.fds[0], F_SETFD, 1) == -1)
		    bailout();
		if (fcntl(nodeptr->out.fds[1], F_SETFD, 1) == -1)
		    bailout();
		if (fcntl(nodeptr->err.fds[0], F_SETFD, 1) == -1)
		    bailout();
		if (fcntl(nodeptr->err.fds[1], F_SETFD, 1) == -1)
		    bailout();
		nodeptr->childpid = fork();
		switch (nodeptr->childpid) {
		    /* its the ol fork and switch routine eh? */
		case -1:
		    bailout();
		    break;
		case 0:
		    /* remove from parent group to avoid kernel
		     * passing signals to children.
		     */
		    (void)setsid();
		    if (piping)
			if (close(STDIN_FILENO) != 0)
			    bailout();
		    if (dup2(nodeptr->out.fds[1], STDOUT_FILENO) 
			!= STDOUT_FILENO) 
			bailout();
		    if (dup2(nodeptr->err.fds[1], STDERR_FILENO)
			!= STDERR_FILENO)
			bailout();
		    if (close(nodeptr->out.fds[0]) != 0)
			bailout();
		    if (close(nodeptr->err.fds[0]) != 0)
			bailout();
		    if (username != NULL)
			(void)snprintf(buf, MAXBUF, "%s@%s", username,
				       nodeptr->name);
		    else
			(void)snprintf(buf, MAXBUF, "%s", nodeptr->name);
		    if (debug)
			    (void)printf("%s %s %s\n", rshstring,
				buf, command);
		    if (testflag && rshport > 0 && porttimeout > 0)
			if (!test_node_connection(rshport, porttimeout,
						  nodeptr))
			    exit(EXIT_SUCCESS);
		    /* handle the script argument */
		    if (script) {
			    scriptbase = basename(scriptname);
			    scriptdir = dirname(scriptname);
			    slen = strlen(rshstring) + strlen(scriptbase)*2 +
				strlen(buf) + strlen(command) + 128;

			    cmd = calloc(4, sizeof(char *));
			    arg = 0;
			    cmd[arg++] = "/bin/sh";
			    cmd[arg++] = "-c";
			    cmd[arg] = (char *)malloc(sizeof(char) * slen);
			    (void)snprintf(cmd[arg], slen,
				"tar cfp - %s | %s %s "
				"'mkdir /tmp/dsh.$$; cd /tmp/dsh.$$ ; "
				"tar xfp - ; ./%s %s; cd /tmp ; "
				"rm -rf /tmp/dsh.$$'", scriptbase,
				rshstring, buf, scriptbase, command);
			    if (debug)
				    (void)printf("%s\n", cmd[arg]);
			    arg++;
			    cmd[arg] = (char *)0;
			    chdir(scriptdir); /* for tar */
			    execvp("/bin/sh", cmd);
			    bailout();
		    }
		    cmd = calloc(nrofargs+1, sizeof(char *));
		    arg = 0;
		    while (rsh[arg] != NULL) {
			    cmd[arg] = rsh[arg];
			    arg++;
		    }
		    cmd[arg++] = buf;
		    cmd[arg++] = command;
		    cmd[arg] = (char *)0;
		    execvp(rsh[0], cmd);
		    bailout();
		    break;
		default:
		    break;
		} /* switch */
		nodeptr = nodeptr->next;
	    } /* for i */
	    nodeptr = nodehold;
	    for (i=0; (i < fanout && nodeptr != NULL); i++) {
		if (gotsigterm)
		    exit(EXIT_FAILURE);
		if (debug)
		    (void)printf("Printing node: %d, fangroup %d,"
				 " fanout part: %d\n", ((n) ? g-fanout+i : i),
				 n, i);
		currentchild = nodeptr->childpid;
		/* now close off the useless stuff, and read the goodies */
		if (close(nodeptr->out.fds[1]) != 0)
		    bailout();
		if (close(nodeptr->err.fds[1]) != 0)
		    bailout();
		fda = fdopen(nodeptr->out.fds[0], "r"); /* stdout */
		if (fda == NULL)
		    bailout();
		fd = fdopen(nodeptr->err.fds[0], "r"); /* stderr */
		if (fd == NULL)
		    bailout();
		fds[0].fd = nodeptr->out.fds[0];
		fds[1].fd = nodeptr->err.fds[0];
		fds[0].events = POLLIN|POLLPRI;
		fds[1].events = POLLIN|POLLPRI;
		pollret = 1;

		while (pollret >= 0) {
		    int gotdata;

		    pollret = poll(fds, 2, 5);
		    gotdata = 0;
		    if ((fds[0].revents&POLLIN) == POLLIN ||
			(fds[0].revents&POLLHUP) == POLLHUP ||
		        (fds[0].revents&POLLPRI) == POLLPRI) {
		    	cd = fgets(pipebuf, sizeof(pipebuf), fda);
		    	if (cd != NULL) {
			    (void)printf("%*s: %s",
			                 -maxnodelen, nodeptr->name, cd);
			    gotdata++;
			}
		    }
		    if ((fds[1].revents&POLLIN) == POLLIN ||
			(fds[0].revents&POLLHUP) == POLLHUP ||
			(fds[1].revents&POLLPRI) == POLLPRI) {
		    	cd = fgets(pipebuf, sizeof(pipebuf), fd);
		    	if (errorflag && cd != NULL) {
			    (void)printf("%*s: %s",
			                 -maxnodelen, nodeptr->name, cd);
			    gotdata++;
			} else if (!errorflag && cd != NULL)
			    gotdata++;
		    }
		    if (!gotdata)
			if (((fds[0].revents&POLLHUP) == POLLHUP ||
			     (fds[0].revents&POLLERR) == POLLERR ||
			     (fds[0].revents&POLLNVAL) == POLLNVAL) &&
			    ((fds[1].revents&POLLHUP) == POLLHUP ||
			     (fds[1].revents&POLLERR) == POLLERR ||
			     (fds[1].revents&POLLNVAL) == POLLNVAL))
			    break;
		}
		fclose(fda);
		fclose(fd);
		(void)wait(&status);
		nodeptr = nodeptr->next;
	    } /* for pipe read */			
	} /* for n */
	if (piping) {
	    /* yes, this is code repetition, no need to adjust your monitor */
	    if (isatty(STDIN_FILENO) && piping)
		(void)printf("%s>", progname);
	    command = fgets(cbuf, sizeof(cbuf), in);
	    if (command != NULL)
		if (strcmp(command,"\n") == 0)
		    command = NULL;
	    nodeptr = nodelink;
	} else
	    command = NULL;
    } /* while loop */
    if (piping) {  /* I learned this the hard way */
	fflush(in);
	fclose(in);
    }
}

void
sig_handler(int i)
{
    switch (i) {
    case SIGINT:
	killpg(currentchild, SIGINT);
	break;
    case SIGTERM:
	gotsigterm = 1;
	break;
    case SIGALRM:
	break;
    default:
	bailout();
	break;
    }
}
